crates/sui-core/src/authority_server.rs

```diff
@@ -1010,6 +1063,136 @@ impl ValidatorService {
         })
     }
 
+    async fn wait_for_effects_impl(
+        &self,
+        request: tonic::Request<RawWaitForEffectsRequest>,
+    ) -> WrappedServiceResponse<RawWaitForEffectsResponse> {
+        let request: WaitForEffectsRequest = request.into_inner().try_into()?;
+        let response = timeout(
+            // TODO(fastpath): Tune this once we have a good estimate of the typical delay.
+            Duration::from_secs(10),
```

Comment:
I think we need to be slightly careful here to not long poll for too long. If this is too long we can end up with 10k+ requests just sitting there. Also do we have any protections for folks waiting on effects that wont ever exist?

Reply:
I think we could do better here.
We have the consensus position in the input, so we could use it to make better decisions.
For instance, if we have far passed the round from the position we can stop waiting. Or if the specified round is too far from tip we could bail.